tools/wake-unit/unit.cpp

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2022 SiFive, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You should have received a copy of LICENSE.Apache2 along with
+ * this software. If not, you may obtain a copy at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "unit.h"
+#include "util/colour.h"
+#include "util/diagnostic.h"
+
+#include <set>
+#include <cstring>
+
+DiagnosticReporter *reporter;
+
+struct Test {
+  const char* test_name;
+  TestFunc test;
+  Test(const char* test_name_, TestFunc test_) : test_name(test_name_), test(test_) {}
+};
+
+// TODO: sequester this off to its own file and make it static
+// so that only TestRegister and the test runner can access it.
+std::vector<Test> tests__;
+
+TestRegister::TestRegister(const char* test_name, TestFunc test) {
+  tests__.emplace_back(test_name, test);
+}
+
+int main(int argc, char** argv) {
+  bool no_color = false;
+  if (argc > 2) {
+    std::cerr << "Too many arguments to be valid, --no-color is the only valid argument";
+    return -1;
+  }
+  if (argc == 2) {
+    if (strcmp(argv[1],"--no-color") == 0) no_color = true;
+    else {
+      std::cerr << "`" << argv[1] << "` is not a valid argument, --no-color is the only valid option";
+      return -1;
+    }
+  }
+  term_init(true);
+  TestLogger logger;
+  std::set<std::string> failed_tests;
+  std::set<std::string> passing_tests;
+  for (auto& test : tests__) {
+    // Set this logjump in case this test fails
+    size_t num_errors = logger.errors.size();
+    logger.test_name = test.test_name;
+    if(setjmp(logger.return_jmp_buffer)) {
```

Comment:
setjmp? really?
Even exceptions have less code smell than setjmp.
Is this just so you can use `assert()`?
Defining your own `wakeAssert()` seems like it would be a better move.

Reply:
It has to be a setjmp though (more on this later) and if you read gtest's docs it actully becomes clear that they're *also* using setjmp since they say they say destructors won't run after an assert.

The reason that setjmp is needed instead of exceptions is because I need to throw in the destructor to get the `<<` assert interface. I was coppying gtest's interface and I'm like 95% sure that this is what they're doing. You have to cause the test function to exit using code only in the destructor of whatever object is returned by an ASSERT_FOO macro, but destructors aren't allowed to throw and you can't return.

An alternative interface would be to now allow the `ASSERT_FOO(...) << "my message here"` interface and to make it something more like `ASSERT_FOO(..., "my message here")` then I can change the macro to use a return/throw instead of a setjmp from a destructor. But if I want that `<<` syntax I have to setjmp, it was unexpected I'm not gonna lie.